AsyncDisplayKit/Layout/ASLayoutSpec.mm

```diff
@@ -42,4 +49,60 @@ - (ASLayout *)measureWithSizeRange:(ASSizeRange)constrainedSize
   return [ASLayout layoutWithLayoutableObject:self size:constrainedSize.min];
 }
 
+- (ASLayoutSpec *)finalLayoutableWithParent:(ASLayoutSpec *)parentSpec;
+{
+  return nil;
+}
+
+- (void)setChild:(id<ASLayoutable>)child;
+{
+  [self setChild:child forIdentifier:kDefaultChildKey];
+}
+
+- (id<ASLayoutable>)layoutableToAddFromLayoutable:(id<ASLayoutable>)child
+{
+  ASLayoutOptions *layoutOptions = [child layoutOptions];
+  
+  id<ASLayoutable> finalLayoutable = [child finalLayoutableWithParent:self];
+  if (finalLayoutable) {
+    [layoutOptions copyIntoOptions:finalLayoutable.layoutOptions];
+    return finalLayoutable;
+  }
+  return child;
+}
+
+- (void)setChild:(id<ASLayoutable>)child forIdentifier:(NSString *)identifier
+{
+  ASDisplayNodeAssert(self.isMutable, @"Cannot set properties when layout spec is not mutable");
+  self.layoutChildren[identifier] = [self layoutableToAddFromLayoutable:child];;
```

Comment:
I think we need assertions to protect against the case of a user calling setChild: and setChildren: on the same spec.  Right now, it would store both, and be a sort of undefined behavior.  It should probably trigger an assertion or exception, and as fallback runtime behavior (e.g. if debug-only assertion), overwrite the last action ("last one wins").

This is fairly important for robustness as a relatively innocent product development mistake could trigger this condition, e.g. imagine experiment-driven code paths that are hard to test.  However, feel free to notate it as a separate followup for a couple weeks from now.